### PR TITLE
fix: git material saved in transaction

### LINKED ISSUE #5045: Bug:
### 📜 Description

When git sensor is down, then we add git materials in orchestrator database and it does not get added in git sensor database. This results in error while configuring branch. In order to fix this, we did the create and update flow of git materials in transaction. If git sensor is down, then the materials should not be saved in database.

### 👟 Reproduction steps

When git sensor is down, then we add git materials in orchestrator database and it does not get added in git sensor database. This results in error while configuring branch. In order to fix this, we did the create and update flow of git materials in transaction. If git sensor is down, then the materials should not be saved in database.

### 👍 Expected behavior

When git sensor is down, then we add git materials in orchestrator database and it does not get added in git sensor database. This results in error while configuring branch. In order to fix this, we did the create and update flow of git materials in transaction. If git sensor is down, then the materials should not be saved in database.

### 👎 Actual Behavior

When git sensor is down, then we add git materials in orchestrator database and it does not get added in git sensor database. This results in error while configuring branch. In order to fix this, we did the create and update flow of git materials in transaction. If git sensor is down, then the materials should not be saved in database.

### ☸ Kubernetes version

AKS

### Cloud provider

<details>
Azure
</details>

### 🌍 Browser

Chrome

### 🧱 Your Environment

_No response_

### ✅ Proposed Solution

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find any similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#9683

### PR DESCRIPTION
**Pull request by ShashwatDadhich**
# Description

This issue was happening when we were saving git meterials in orchestrator and not in git sensor

Fixes #5045

## How Has This Been Tested?

- [x] Tested locally 
- [x] Tested by deploying in VM.

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR requires documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [x] I have added all the required unit/api test cases.

### CODE REVIEW

File: pkg/pipeline/CiCdPipelineOrchestrator.go [hunk 1]
```diff
@@ -1271,6 +1273,18 @@ func (impl CiCdPipelineOrchestratorImpl) DeleteApp(appId int, userId int32) erro
...
 func (impl CiCdPipelineOrchestratorImpl) CreateMaterials(createMaterialRequest *bean.CreateMaterialDTO) (*bean.CreateMaterialDTO, error) {
+	tx, err := impl.transactionManager.StartTx()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			err := tx.Rollback()
+			if err != nil {
+				impl.logger.Errorw("error in rollback Create material", "err", err)
```
Comment (iamayushm) [inline pkg/pipeline/CiCdPipelineOrchestrator.go:1284]: log metadata

File: pkg/pipeline/CiCdPipelineOrchestrator.go [hunk 2]
```diff
@@ -1295,7 +1309,7 @@ func (impl CiCdPipelineOrchestratorImpl) CreateMaterials(createMaterialRequest *
 	var materials []*bean.GitMaterial
 	for _, inputMaterial := range createMaterialRequest.Material {
 		inputMaterial.UpdateSanitisedGitRepoUrl()
-		m, err := impl.createMaterial(inputMaterial, createMaterialRequest.AppId, createMaterialRequest.UserId)
+		m, err := impl.createMaterial(inputMaterial, createMaterialRequest.AppId, createMaterialRequest.UserId, tx)
```
Comment (iamayushm) [inline pkg/pipeline/CiCdPipelineOrchestrator.go:1312]: pass tx as first parameter

File: pkg/pipeline/CiCdPipelineOrchestrator.go [hunk 3]
```diff
@@ -1307,12 +1321,29 @@ func (impl CiCdPipelineOrchestratorImpl) CreateMaterials(createMaterialRequest *
 		impl.logger.Errorw("error in updating to sensor", "err", err)
 		return nil, err
 	}
+	err = impl.transactionManager.CommitTx(tx)
+	if err != nil {
+		impl.logger.Errorw("error in committing tx Create material", "err", err)
```
Comment (iamayushm) [inline pkg/pipeline/CiCdPipelineOrchestrator.go:1326]: log metadata

File: pkg/pipeline/CiCdPipelineOrchestrator.go [hunk 4]
```diff
@@ -1307,12 +1321,29 @@ func (impl CiCdPipelineOrchestratorImpl) CreateMaterials(createMaterialRequest *
...
 }
 
 func (impl CiCdPipelineOrchestratorImpl) UpdateMaterial(updateMaterialDTO *bean.UpdateMaterialDTO) (*bean.UpdateMaterialDTO, error) {
-	updatedMaterial, err := impl.updateMaterial(updateMaterialDTO)
+	tx, err := impl.transactionManager.StartTx()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
```
Comment (iamayushm) [inline pkg/pipeline/CiCdPipelineOrchestrator.go:1339]: wrong usage of error variable, review

File: pkg/pipeline/history/repository/GitMaterialHistoryRepository.go [hunk 5]
```diff
@@ -21,7 +21,7 @@ type GitMaterialHistory struct {
 }
 
 type GitMaterialHistoryRepository interface {
-	SaveGitMaterialHistory(material *GitMaterialHistory) error
+	SaveGitMaterialHistory(material *GitMaterialHistory, tx *pg.Tx) error
```
Comment (iamayushm) [inline pkg/pipeline/history/repository/GitMaterialHistoryRepository.go:24]: use tx as first input param everywhere

**Review by iamayushm — APPROVED** [review]

**Review by iamayushm — APPROVED** [review]

**Review by nishant-d — APPROVED** [review]

**Pull request merged**